llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -231,6 +231,48 @@ static KernelMapEntryScope selectDeviceCodeSplitScopeAutomatically(Module &M) {
   return Scope_PerModule;
 }
 
+// Collect all the dependencies for the function.
+static void collectFunctionCallGraphNodes(llvm::Function *Func,
+                                          SetVector<const GlobalValue *> &GVs) {
+  if (!GVs.count(Func))
+    GVs.insert(Func);
+
+  std::vector<llvm::Function *> Workqueue;
+  Workqueue.push_back(Func);
+
+  while (!Workqueue.empty()) {
+    Function *F = &*Workqueue.back();
+    Workqueue.pop_back();
+    for (auto &I : instructions(F)) {
+      if (CallBase *CB = dyn_cast<CallBase>(&I))
+        if (Function *CF = CB->getCalledFunction())
+          if (!CF->isDeclaration() && !GVs.count(CF)) {
+            GVs.insert(CF);
+            Workqueue.push_back(CF);
+          }
+    }
+  }
+}
+
+// This function returns the collection of functions that are not
+// called from any other function in the module. Example of such
+// an unreferenced function can be a SYCL_EXTERNAL function that
+// will be referenced later, e.g. during linking SPIRV modules.
+static std::vector<llvm::Function *> collectUnreferencedFuncs(Module &M) {
+  SetVector<const GlobalValue *> ReferencedFuncs;
+  for (auto &F : M.functions()) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL)
+      collectFunctionCallGraphNodes(&F, ReferencedFuncs);
+  }
+
+  std::vector<llvm::Function *> UnreferencedFuncs;
+  for (auto &F : M.functions()) {
+    if (!F.isDeclaration() && !ReferencedFuncs.count(&F))
+      UnreferencedFuncs.push_back(&F);
+  }
```

Comment:
Is it possible to achieve the same in some easier way? Like for example we could search for uses of function and if there are no calls to it, then it is considered as unreferenced.